File: src/log.cpp
#include "log.hpp"

/**
 * @brief Constructor for Log class that initializes the log file and log level.
 *
 * @param config Shared pointer to configuration settings which include log file
 * path and level.
 */
Log::Log(const std::shared_ptr<Config>& config)
    : config_(config), level_(Level::INFO) {
  FUCK(config_->modeToString());
  if (config_->runMode() == RunMode::server)
    mode_ = std::string("SERVER");
  else if (config_->runMode() == RunMode::agent)
    mode_ = std::string("AGENT");
  // Open the log file for appending.
  std::ofstream logFile(config_->log().file, std::ios::out | std::ios::app);
  if (!logFile.is_open()) {
    // Print an error message if the log file cannot be opened.
    std::cerr << "Error opening log file: " << config_->log().file
              << ". Make sure the directory and file exist.\n";
  }

  // Set the log level based on the configuration.
  const auto& logLevel = config_->log().level;
  if (logLevel == "INFO") {
    level_ = Level::INFO;
  } else if (logLevel == "TRACE") {
    level_ = Level::TRACE;
  } else if (logLevel == "DEBUG") {
    level_ = Level::DEBUG;
  } else {
    // Print an error message if the log level is invalid.
    std::cerr << "Invalid log level: " << logLevel
              << ". It should be one of [INFO|TRACE|DEBUG].\n";
  }
}

/**
 * @brief Copy constructor for Log class.
 *
 * @param log Shared pointer to another Log instance to copy from.
 */
Log::Log(const std::shared_ptr<Log>& log)
    : config_(log->config_), level_(log->level_) {}

Log::~Log() {}  ///< Destructor for Log class.

/**
 * @brief Write a log message to the log file and console.
 *
 * @param message The log message to write.
 * @param level The log level of the message.
 */
void Log::write(const std::string& message, Level level) const {
  // Check if the message should be logged based on its level.
  if (level <= level_ || level == Level::ERROR) {
    // Open the log file for appending.
    std::ofstream logFile(config_->log().file, std::ios::out | std::ios::app);
    if (logFile.is_open()) {
      // Get the current time and format it as a timestamp.
      auto now = std::time(nullptr);
      auto localTime = *std::localtime(&now);
      std::ostringstream timestampStream;
      timestampStream << std::put_time(&localTime, "%Y-%m-%d_%H:%M:%S");
      std::string timestamp = timestampStream.str();

      // Create a formatted log line.
      std::string line = timestamp + " [" + mode_ + "]" + " [" +
                         levelToString(level) + "] " + message + "\n";
      logFile << line;    // Write the line to the log file.
      std::cout << line;  // Also write the line to the console.
    } else {
      // Print an error message if the log file cannot be opened.
      std::cerr << "Error opening log file: " << config_->log().file
                << ". Make sure the directory and file exist.\n";
    }
  }
}

/**
 * @brief Convert log level enumeration to its string representation.
 *
 * @param level The log level to convert.
 * @return String representation of the log level.
 */
std::string Log::levelToString(Level level) {
  switch (level) {
    case Level::INFO:
      return "INFO";
    case Level::TRACE:
      return "TRACE";
    case Level::ERROR:
      return "ERROR";
    case Level::DEBUG:
      return "DEBUG";
    default:
      return "UNKNOWN";  // Return "UNKNOWN" for any unrecognized log level.
  }
}

File: src/log.rs
//! Logging. Ports `log.{hpp,cpp}`.
//!
//! Level gating matches C++: a message is emitted when `level <= configured`
//! (by the enum ordering below) OR the message is `ERROR`. Output goes to both
//! the log file (append) and stdout, formatted `{ts} [MODE] [LEVEL] msg`.

use std::fs::{File, OpenOptions};
use std::io::Write;
use std::sync::Mutex;

use crate::config::{Config, RunMode};

/// Order matches the C++ `Log::Level` enum (INFO=0, TRACE=1, ERROR=2, DEBUG=3).
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum Level {
    Info = 0,
    Trace = 1,
    Error = 2,
    Debug = 3,
}

impl Level {
    fn as_str(self) -> &'static str {
        match self {
            Level::Info => "INFO",
            Level::Trace => "TRACE",
            Level::Error => "ERROR",
            Level::Debug => "DEBUG",
        }
    }

    fn from_config(s: &str) -> Self {
        match s {
            "TRACE" => Level::Trace,
            "DEBUG" => Level::Debug,
            "INFO" => Level::Info,
            other => {
                eprintln!("Invalid log level: {other}. It should be one of [INFO|TRACE|DEBUG].");
                Level::Info
            }
        }
    }
}

pub struct Log {
    level: Level,
    mode: &'static str,
    file_path: String,
    file: Mutex<Option<File>>,
}

impl Log {
    pub fn new(config: &Config) -> Self {
        let mode = match config.run_mode {
            RunMode::Server => "SERVER",
            RunMode::Agent => "AGENT",
        };

        let file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&config.log.file);

        let handle = match file {
            Ok(f) => Some(f),
            Err(e) => {
                eprintln!(
                    "Error opening log file: {}. Make sure the directory and file exist. ({e})",
                    config.log.file
                );
                None
            }
        };

        Log {
            level: Level::from_config(&config.log.level),
            mode,
            file_path: config.log.file.clone(),
            file: Mutex::new(handle),
        }
    }

    pub fn write(&self, message: &str, level: Level) {
        if !(level <= self.level || level == Level::Error) {
            return;
        }

        let ts = chrono::Local::now().format("%Y-%m-%d_%H:%M:%S");
        let line = format!("{ts} [{}] [{}] {message}\n", self.mode, level.as_str());

        print!("{line}");
        let _ = std::io::stdout().flush();

        let mut guard = self.file.lock().unwrap();
        if guard.is_none() {
            *guard = OpenOptions::new()
                .create(true)
                .append(true)
                .open(&self.file_path)
                .ok();
        }
        if let Some(f) = guard.as_mut() {
            let _ = f.write_all(line.as_bytes());
        } else {
            eprintln!(
                "Error opening log file: {}. Make sure the directory and file exist.",
                self.file_path
            );
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn gating_info_level() {
        // configured=INFO shows INFO + ERROR only
        assert!(Level::Info <= Level::Info);
        assert!(!(Level::Trace <= Level::Info));
        assert!(!(Level::Debug <= Level::Info));
        assert_eq!(Level::Error, Level::Error);
    }

    #[test]
    fn gating_debug_shows_all() {
        for l in [Level::Info, Level::Trace, Level::Error, Level::Debug] {
            assert!(l <= Level::Debug || l == Level::Error);
        }
    }
}
